hoomd/mpcd/ParticleData.cc

```diff
@@ -416,6 +457,128 @@ void mpcd::ParticleData::initializeRandom(unsigned int N, const BoxDim& local_bo
+    std::uniform_real_distribution<Scalar> pos_z(lo.z, hi.z);
+    std::normal_distribution<Scalar> vel(0.0, fast::sqrt(kT / m_mass));
+
+    // generating Ndraw random positions and keeping them if they lie inside sphere(also counting N_kept)
+    unsigned int N_kept = 0;
+    unsigned int N_tag_rank = 0;
+
+    for (unsigned int i=0; i < Ndraw; ++i)
+        {
+        Scalar pos_x_draw = pos_x(mt);
+        Scalar pos_y_draw = pos_y(mt);
+        Scalar pos_z_draw = (ndimensions == 3) ? pos_z(mt) : Scalar(0.0);
+        Scalar mod_pos_draw = slow::sqrt(pos_x_draw*pos_x_draw + pos_y_draw*pos_y_draw + pos_z_draw*pos_z_draw);
+        if (mod_pos_draw < R)
+            {
+            h_pos_draw.data[N_kept] =  make_scalar4(pos_x_draw,
+                                                    pos_y_draw,
+                                                    pos_z_draw,
+                                                    __int_as_scalar(0));
+            N_kept += 1;
+            }
+        }
+
+    // setting m_N to N_kept(actual drawn particles) on that rank
+    m_N = N_kept;
+    
+    // reduce / scan the number of particles that are drawn on all ranks to get total N and tag on current rank
+    #ifdef ENABLE_MPI
+    const unsigned int nrank = m_exec_conf->getNRanks();
+    if (nrank > 1)
+        {
+        MPI_Allreduce(MPI_IN_PLACE, &N_kept, 1, MPI_UNSIGNED, MPI_SUM, m_exec_conf->getMPICommunicator());
+        MPI_Exscan(&N_kept, &N_tag_rank, 1, MPI_UNSIGNED, MPI_SUM, m_exec_conf->getMPICommunicator());
+        }
+    #endif
+    
+    // setting N_global
+    unsigned int N = N_kept;
+    setNGlobal(N);
+    //allocate and fill up with random values generated earlier
+    allocate(m_N);
+    ArrayHandle<Scalar4> h_pos(m_pos, access_location::host, access_mode::overwrite);
+    ArrayHandle<Scalar4> h_vel(m_vel, access_location::host, access_mode::overwrite);
+    ArrayHandle<unsigned int> h_tag(m_tag, access_location::host, access_mode::overwrite);
+    double3 vel_cm = make_double3(0,0,0);
+
+    for (unsigned int i=0; i < m_N; ++i)
+        {
+        h_pos.data[i] = h_pos_draw.data[i];
+        h_vel.data[i] = make_scalar4(vel(mt),
+                                     vel(mt),
+                                     (ndimensions == 3) ? vel(mt) : Scalar(0.0),
+                                     __int_as_scalar(mpcd::detail::NO_CELL));
+        h_tag.data[i] = N_tag_rank + i ;
+
+        // add up total velocity
+        vel_cm.x += h_vel.data[i].x;
+        vel_cm.y += h_vel.data[i].y;
+        vel_cm.z += h_vel.data[i].z;
+        }
+
+    // compute average velocity per-particle to remove
+    #ifdef ENABLE_MPI
+    if (nrank > 1)
+        {
+        MPI_Allreduce(MPI_IN_PLACE, &vel_cm, 3, MPI_DOUBLE, MPI_SUM, m_exec_conf->getMPICommunicator());
+        }
+    #endif // ENABLE_MPI
+    if (N > 0)
+        {
+        vel_cm.x /= N;
+        vel_cm.y /= N;
+        vel_cm.z /= N;
+        }
```

Comment:
I know this was `N` in the code above, but using `m_N_global` is clearer about what's going on:
```suggestion
    if (m_N_global > 0)
        {
        vel_cm.x /= m_N_global;
        vel_cm.y /= m_N_global;
        vel_cm.z /= m_N_global;
        }
```